### PR TITLE
Enable hash algorithm selection

### PR DESCRIPTION
**Pull request by KevinRansom**
1.  Add a switch to the compiler to enable hashcode selection {SHA1|SHA256}.
     Defaults to SHA256, if none specified will use SHA256.  It can't be switched off.
2.  Updated command line tests.
3.  Update Build targets and BuildTask to enable configuration of hash code.
     Use <PdbChecksumAlgorithm>SHA1</PdbChecksumAlgorithm> in .fsproj to switch to SHA1 hash.
4.  Use the selected when generating the hash for portable PDBs
5.  Write Hash alg+hash to Debug_DIRECTORY of the PE file.
6.  Deterministic didn't write a record in the to Debug_DIRECTORY of the PE file
7.  Always create a deterministic PDB, because we hash anyway.  So Id, always fits the contents of the PDB file. 
8.  Fixed the Deterministic Id providers hashing.

I think I need to do some refactoring around pe generation and the pdb writing.  I will take a look at it.

### CODE REVIEW

**Comment by cartermp** [comment]
@KevinRansom @tmat is there a tracking issue for this? I recall seeing it somewhere...

**Comment by KevinRansom** [comment]
Fixes:
- https://github.com/dotnet/fsharp/issues/1223
- https://github.com/dotnet/fsharp/issues/4259

**Comment by KevinRansom** [comment]
@dsyme can you run your eyes over this mate.  It's needed for 16.3.

Thanks

Kevin

**Comment by KevinRansom** [comment]
@dsyme can you run your eyes over this mate.  It's needed for 16.3.

Thanks

Kevin

File: src/absil/ilwritepdb.fs [hunk 1]
```diff
@@ -125,6 +126,30 @@ type idd =
       iddData: byte[]
       iddChunk: BinaryChunk }
 
+/// The specified Hash algorithm to use on portable pdb files.
+type HashAlgorithm =
+| Sha1
```
Comment (cartermp) [inline src/absil/ilwritepdb.fs]: indent 4 spaces

File: src/absil/ilwritepdb.fs [hunk 2]
```diff
@@ -125,6 +126,30 @@ type idd =
...
+| Sha256
```
Comment (cartermp) [inline src/absil/ilwritepdb.fs]: Same

File: src/absil/ilwritepdb.fs [hunk 3]
```diff
@@ -125,6 +126,30 @@ type idd =
...
+
+// Document checksum algorithms
+let guidSha1 = Guid("ff1816ec-aa5e-4d10-87f7-6f4963833460")
+let guidSha2 = Guid("8829d00f-11b8-4213-878b-770e8597ac16")
+
+// If the FIPS algorithm policy is enabled on the computer (e.g., for US government employees and contractors)
```
Comment (cartermp) [inline src/absil/ilwritepdb.fs]: This comment looks out of date since MD5 is not being used.

File: src/absil/ilwritepdb.fsi [hunk 4]
```diff
@@ -83,10 +83,14 @@ type idd =
       iddData: byte[];
       iddChunk: BinaryChunk }
 
-val generatePortablePdb : embedAllSource:bool -> embedSourceList:string list -> sourceLink: string -> showTimes:bool -> info:PdbData -> isDeterministic:bool -> pathMap:PathMap -> (int64 * BlobContentId * MemoryStream)
+type HashAlgorithm =
+| Sha1
```
Comment (cartermp) [inline src/absil/ilwritepdb.fsi]: Indent 4 spaces

File: src/absil/ilwritepdb.fsi [hunk 5]
```diff
@@ -83,10 +83,14 @@ type idd =
...
+| Sha256
```
Comment (cartermp) [inline src/absil/ilwritepdb.fsi]: Same

File: src/absil/ilwritepdb.fs (hunk 2 above)
Comment (KevinRansom) [reply src/absil/ilwritepdb.fs]: Sigh!!! I prefer aligning enums under the type.

File: src/absil/ilwritepdb.fs [hunk 6]
```diff
@@ -169,28 +189,52 @@ let pdbGetPdbDebugInfo (embeddedPDBChunk: BinaryChunk) (uncompressedLength: int6
...
+      iddChunk = checksumPdbChunk
+    }
 
-// Document checksum algorithms
-let guidSourceHashMD5 = System.Guid(0x406ea660u, 0x64cfus, 0x4c82us, 0xb6uy, 0xf0uy, 0x42uy, 0xd4uy, 0x81uy, 0x72uy, 0xa7uy, 0x99uy) //406ea660-64cf-4c82-b6f0-42d48172a799
-let hashSizeOfMD5 = 16
+let pdbGetPdbDebugDeterministicInfo (deterministicPdbChunk: BinaryChunk) =
+    { iddCharacteristics = 0                                                    // Reserved
+      iddMajorVersion = 0                                                       // VersionMajor should be 0
+      iddMinorVersion = 0                                                       // VersionMinor should be 0x0100
```
Comment (brettfo) [inline src/absil/ilwritepdb.fs]: Comment and data don't agree.

Comment (KevinRansom) [reply src/absil/ilwritepdb.fs]: > Comment and data don't agree.

Indeed they don't code is right, comment is copy/paste error.

Thx.

**Pull request merged**